### PR TITLE
A simple TreeMapBuilder

### PR DESCRIPTION
**Pull request by mkeskells**
Defer the creation of the TreeMap until be expose it

Don't create tuples if we don't need to, as we can use foreachEntry

### CODE REVIEW

**Comment by mkeskells** [comment]
Mima issues
```
[error]  * method this(scala.collection.immutable.RedBlackTree#Tree,scala.math.Ordering)Unit in class scala.collection.immutable.TreeMap does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.TreeMap.this")
[error]  * object scala.collection.immutable.TreeMap#TreeMapBuilder#adder does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.TreeMap$TreeMapBuilder$adder$")
[error]  * class scala.collection.immutable.TreeMap#TreeMapBuilder does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.TreeMap$TreeMapBuilder")
```

**Review by diesalbla — APPROVED** [review]

File: src/library/scala/collection/immutable/TreeMap.scala [hunk 1]
```diff
@@ -27,6 +27,67 @@ object TreeMap extends ImmutableSortedMapFactory[TreeMap] {
...
+      }
+
+      override def apply(key: A, value: B): Unit = {
+        accumulator = RB.update(accumulator, key, value, overwrite = true)
+      }
+    }
+
+    override def ++=(xs: TraversableOnce[(A, B)]): this.type = {
+      xs match {
+        case ts: TreeMap[A, B] if ts.ordering eq ordering =>
```
Comment (retronym) [inline src/library/scala/collection/immutable/TreeMap.scala]: Our optimizations elsewhere are more permissive and allow use `==` when comparing `Ordering`-s. Let's review the rest of the code to see if that is consistently done, and if so, also do so here.

Comment (mkeskells) [reply src/library/scala/collection/immutable/TreeMap.scala]: it seems to be consistently used as `==` in `TreeMap` and `TreeSet` so updated here

File: src/library/scala/collection/immutable/TreeMap.scala [hunk 2]
```diff
@@ -27,6 +27,67 @@ object TreeMap extends ImmutableSortedMapFactory[TreeMap] {
...
+      }
+    }
+
+    override def ++=(xs: TraversableOnce[(A, B)]): this.type = {
+      xs match {
+        case ts: TreeMap[A, B] if ts.ordering == ordering =>
+          if (tree eq null) tree = ts.tree0
+          else tree = adder.addTree(tree, ts.tree0)
+        case that: HasForeachEntry[A, B] =>
+          tree = adder.addForEach(tree, that)
```
Comment (retronym) [inline src/library/scala/collection/immutable/TreeMap.scala:78]: (I think this you have this planned for a follow PR, but noting here anyway)

We could add a special case when `xs: RedBlackTree[A, B] if xs.ordering == ordering` to directly assign its node to `this.tree` when `tree` is null (ie it is the first thing seen by the builder), or to use a bulk update using mutation.

**Review by retronym — APPROVED** [review]

**Pull request merged**